### PR TITLE
Simulator is Complete!! 🚀🚀

### PR DESCRIPTION
**Pull request by JeremyButt**
Completed an operational simulator for simulating chirpnets devices. This will facilitate the development of the routing algorithm and packet design.

Hit me up if you have any questions! README is enroute.

### CODE REVIEW

**Comment by s-axford** [comment]
This is looking great! This will speed up our testing of any routing exponentially!

So the intention is to have the function in SimulatorLib be interchangeable with the true embedded functions? 

I noticed you have a native interface called main() in the VirtualDeviceWrapper, is this where the setup() and then loop() will be called? If so we can take the setup() and loop() I have written as a starting point and start integrating the routing algorithm!

**Comment by johnnyb8** [comment]
This is a big boy. Good work @JeremyButt :)

**Comment by JeremyButt** [comment]
> will be called

**Comment by JeremyButt** [comment]
Didn't mean to close this... im making some changes though and added a readme. will be back up in a bit!

**Comment by JeremyButt** [comment]
Didn't mean to close this... im making some changes though and added a readme. will be back up in a bit!

**Comment by JeremyButt** [comment]
Merge This when everyone is good with it! :)

File: simulator/pom.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>RELEASE</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+
+</project>
```
Comment (TimothyMather) [inline simulator/pom.xml]: No new line, I think you can add into settings to add a newline on save.

File: simulator/src/main/cpp/SimulatorLib.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    jbyte *bytes;
+    bytes = env->GetByteArrayElements(array, 0);
+    int chars_len = env->GetArrayLength(array);
+    chars = new char[chars_len + 1];
+    memset(chars,0,chars_len + 1);
+    memcpy(chars, bytes, chars_len);
+    chars[chars_len] = 0;
+    env->ReleaseByteArrayElements(array, bytes, 0);
+    return chars;
+}
```
Comment (TimothyMather) [inline simulator/src/main/cpp/SimulatorLib.cpp]: Here too

**Comment by TimothyMather** [comment]
I'd say take a look for newlines at the end of files I saw 2 or 3. Other than that its good for me!

**Comment by s-axford** [comment]
I had acouple questions in my comment above, but once those are answered I am good with it

**Comment by JeremyButt** [comment]
> So the intention is to have the function in SimulatorLib be interchangeable with the true embedded functions?

Yeah that is the idea. I can make changes to it if required. I did it this way because the current hardware interfaces are separated into a tx and rx where as the powerlabs lib is a single tx/rx interface. i think we should write a wrapper to the powerlabs lib that will mimic the simulationlib here. This will enable the loop and setup function to abstract away from the library being used. ie the sim lib or the powerlabs lib.

> I noticed you have a native interface called main() in the VirtualDeviceWrapper, is this where the setup() and then loop() will be called? If so we can take the setup() and loop() I have written as a starting point and start integrating the routing algorithm!

The main() is the loop class. With the java stuff there is no setup required. i can change the name if required. 

 NOTE : im super buzzed rn so hit up slack when im sober with questions.

**Comment by JeremyButt** [comment]
@DelightedWaif ill add the newlines

File: simulator/pom.xml (hunk 1 above)
Comment (JeremyButt) [reply simulator/pom.xml]: will do

**Comment by JeremyButt** [comment]
@s-axford @DelightedWaif Let me know if there are any other changes.

**Comment by JeremyButt** [comment]
Merge if there are no other changes

**Pull request merged**